src/Filesystem.cpp

```diff
@@ -417,29 +322,178 @@ std::string Filesystem::workingPath(const std::string& filename) const
 
-	// This is a naive approach but works for most cases.
-	size_t pos = path.find_last_of(".");
+    auto p = FS::path{ path };
+    if(!p.has_extension()) {
+        if(mVerbose) { std::cout << "Filesystem::extension(): File '" << path << "' has no extension." << std::endl; }
+        return std::string();
+    }
 
-	if (pos != std::string::npos)
-	{
-		return path.substr(pos + 1);
-	}
-	else if (isDirectory(path))
-	{
-		if (mVerbose) { std::cout << "Filesystem::extension(): Given path '" << path << "' is a directory, not a file." << std::endl; }
-		return std::string();
-	}
-	else
-	{
-		if (mVerbose) { std::cout << "Filesystem::extension(): File '" << path << "' has no extension." << std::endl; }
-		return std::string();
-	}
+    if(FS::is_directory(p)) {
+        if(mVerbose) { std::cout << "Filesystem::extension(): Given path '" << path << "' is a directory, not a file." << std::endl; }
+        return std::string();
+    }
+
+    return p.extension().string();
 }
 
+void Filesystem::setWorkingDirectory(const std::filesystem::path& p)
+{
+    try {
+        std::filesystem::current_path(p);
+        mWorkingDirectory = p;
+    } catch(const std::filesystem::filesystem_error& e) {
+        std::cout << "Filesystem::extension(): std::filesystem reported an error:"
+            << "\n  What: " << e.what()
+            << "\n  Code: " << e.code()
+            << "\n  Path1: \"" << e.path1() << '\"'
+            << "\n  Path2: \"" << e.path2() << '\"'
+            << "\nRethrowing as NAS2D::Exception::filesystem_file_not_found";
+        std::cout.flush();
+        std::string desc{"Path 1: " + e.path1().string() + "Path 2: " + e.path2().string() };
+        throw NAS2D::Exception::filesystem_file_not_found(desc);
+    }
+}
 
-const char* Filesystem::getLastPhysfsError() const
+
+std::filesystem::path Filesystem::getWorkingDirectory()
 {
-	return PHYSFS_getErrorByCode(PHYSFS_getLastErrorCode());
+    //Update the cache if it isn't correct.
+    const auto& cur_path = std::filesystem::current_path();
+    if(mWorkingDirectory != cur_path) {
+        setWorkingDirectory(cur_path);
+        mWorkingDirectory = cur_path;
+    }
+    return mWorkingDirectory;
+}
+
+bool Filesystem::isInit() const
+{
+    return mIsInit;
+}
+
+bool Filesystem::exists(const std::filesystem::path& p) const
+{
+    return std::filesystem::exists(p);
+}
+
+void Filesystem::forEachFileInFolder(const std::filesystem::path& folderpath, const std::string& validExtensionList /*= std::string{}*/, const std::function<void(const std::filesystem::path&)>& callback /*= [](const std::filesystem::path& p) { (void)p; }*/, bool recursive /*= false*/) const {
+    namespace FS = std::filesystem;
+    auto preferred_folderpath = folderpath;
+    preferred_folderpath.make_preferred();
+    bool exists = FS::exists(preferred_folderpath);
+    bool is_directory = FS::is_directory(preferred_folderpath);
+    bool is_folder = exists && is_directory;
+    if(!is_folder) {
```

Comment:
Condense into one line

Reply:
This was a "readability over terse code" choice. Much easier to understand what the code is doing, but I'll change it.

Reply:
Changed if-statement to single line.